### PR TITLE
Added TransposedArray.

### PR DESCRIPTION
**Pull request by pelson**

### CODE REVIEW

**Comment by rhattersley** [comment]
Superb! Thanks @pelson :smile:

File: biggus/tests/unit/test_TransposedArray.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    def test_too_many_dims(self):
+        msg = 'length 5, but should be of length 4'
+        with self.assertRaisesRegexp(ValueError, msg):
+            self.a._apply_axes_mapping(list('abcde'))
+
+
+class Test_shape(unittest.TestCase):
+    def setUp(self):
+        self.zeros = ConstantArray([10, 20, 40, 50])
+        self.a = TransposedArray(self.zeros, [1, 3, 0, 2])
```
Comment (rhattersley) [inline biggus/tests/unit/test_TransposedArray.py]: I think you can ditch this `setUp`.

**Comment by rhattersley** [comment]
The test failures will be fixed by #118.

**Comment by pelson** [comment]
Thanks @rhattersley. I know some of the implementation wasn't as pretty as it could be so :+1: for going for pace over perfection.

Also, it may be that the ``_full_slices`` function may be helpful for #116 (e.g. such that sliced_shape can deal with Ellipsis correctly [for numpy]).

File: biggus/__init__.py [hunk 2]
```diff
@@ -936,6 +936,131 @@ def _groups_of(length, total_length):
...
+    @property
+    def shape(self):
+        return self._apply_axes_mapping(self.pre_transposed.shape)
+
+    def __getitem__(self, keys):
+        keys = _full_keys(keys, self.ndim)
+        new_transpose_order = self.axes[:]
+
+        # Split the keys into np.newaxis and normal keys.
+        new_axes = [axis for axis, key in enumerate(keys) if key is None]
```
Comment (pelson) [inline biggus/__init__.py]: I'm sure this didn't pass you by @rhattersley - TransposedArrays can be indexed with ``np.newaxis``. Essentially, if you were desperate to index with newaxis, is *would* be possible to do:

```
TransposedArray(lazy_array, range(lazy_array.ndim))[..., np.newaxis, :, np.newaxis]
```

I'm not saying that is a thing you'd want to do, but it is a fallback option (and hopefully shows that ``_full_keys`` is helpful for interpreting the keys passed to ``__getitem__``).

**Pull request merged**